### DISCUSSION TITLE
lemmy ports

### DISCUSSION THREAD

**Discussion by mcmikemn** [EnigmaCurry/d.rymcg.tech#91] [Q&A]
I'm trying to add Lemmy (`lemmy` branch). The lemmy service is failing to connect to the postgres service. It was connecting before I turned the nginx ("proxy") service on, so I think it has something to do with ports now that both nginx and Traefik are reverse-proxying. The lemmy log says:
```
thread 'main' panicked at 'Error connecting to postgres://lemmy:86ruE60ExlfUQwzAS7e6IGqOB8Z93JkP@postgres:5432
/lemmy: could not connect to server: Connection refused
lemmy-lemmy-1     |     Is the server running on host "postgres" (192.168.208.4) and accepting
lemmy-lemmy-1     |     TCP/IP connections on port 5432?
```

But `docker inspect lemmy-postgres-1` shows:
```
"Ports": {
"5432/tcp": null
...
"IPAddress": "192.168.208.4",
```

**Comment by mcmikemn** [comment]
I can ping `postgres` from the `lemmy` container.

**Reply by EnigmaCurry** [reply]
- they are on the same default network
- lemmy depends on postgres

so it should work, I just wonder if the postgres container failed?

**Reply by EnigmaCurry** [reply]
if the postgres container is up and functional, try restarting the lemmy container, if it then works it might be a timing issue that postgres isn't starting fast enough, and lemmy tries to connect before its ready.

**Reply by mcmikemn** [reply]
> so it should work, I just wonder if the postgres container failed?

The postgres log has some "weird" stuff in it - like it shuts down and restarts a couple times, but maybe that's normal? The log ends with `database system is ready to accept connections`.
```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... UTC
creating configuration files ... ok
running bootstrap script ... ok
sh: locale: not found
2023-07-30 21:07:12.964 UTC [30] WARNING:  no usable system locales were found
performing post-bootstrap initialization ... ok
syncing data to disk ... ok


Success. You can now start the database server using:

pg_ctl -D /var/lib/postgresql/data -l logfile start

initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
waiting for server to start....2023-07-30 21:07:15.310 UTC [36] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, 64-bit
2023-07-30 21:07:15.331 UTC [36] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-07-30 21:07:15.340 UTC [39] LOG:  database system was shut down at 2023-07-30 21:07:14 UTC
2023-07-30 21:07:15.362 UTC [36] LOG:  database system is ready to accept connections
done
server started
CREATE DATABASE


/usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*

waiting for server to shut down....2023-07-30 21:07:15.620 UTC [36] LOG:  received fast shutdown request
2023-07-30 21:07:15.627 UTC [36] LOG:  aborting any active transactions
2023-07-30 21:07:15.637 UTC [36] LOG:  background worker "logical replication launcher" (PID 42) exited with exit code 1
2023-07-30 21:07:15.643 UTC [37] LOG:  shutting down
2023-07-30 21:07:15.654 UTC [37] LOG:  checkpoint starting: shutdown immediate
2023-07-30 21:07:15.732 UTC [37] LOG:  checkpoint complete: wrote 918 buffers (5.6%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.053 s, sync=0.022 s, total=0.090 s; sync files=250, longest=0.009 s, average=0.001 s; distance=4222 kB, estimate=4222 kB
2023-07-30 21:07:15.747 UTC [36] LOG:  database system is shut down
done
server stopped

PostgreSQL init process complete; ready for start up.

2023-07-30 21:07:15.864 UTC [1] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, 64-bit
2023-07-30 21:07:15.872 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-07-30 21:07:15.874 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2023-07-30 21:07:15.880 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-07-30 21:07:15.892 UTC [52] LOG:  database system was shut down at 2023-07-30 21:07:15 UTC
2023-07-30 21:07:15.899 UTC [1] LOG:  database system is ready to accept connections
2023-07-30 21:12:16.003 UTC [50] LOG:  checkpoint starting: time
2023-07-30 21:13:45.939 UTC [50] LOG:  checkpoint complete: wrote 899 buffers (5.5%); 0 WAL file(s) added, 0 removed, 1 recycled; write=89.911 s, sync=0.010 s, total=89.937 s; sync files=228, longest=0.002 s, average=0.001 s; distance=10776 kB, estimate=10776 kB
```


> if the postgres container is up and functional, try restarting the lemmy container, if it then works it might be a timing issue that postgres isn't starting fast enough, and lemmy tries to connect before its ready.

I tried restarting the lemmy container, same result. Strangely, its log has the "error connecting to postgres" in it, but ends with `federation enabled, host is lemmy.d.thewooskeys.com. Starting http server at 0.0.0.0:8536`
```
thread 'main' panicked at 'Error connecting to postgres://lemmy:86ruE60ExlfUQwzAS7e6IGqOB8Z93JkP@postgres:5432/lemmy: could not connect to server: Connection refused
Is the server running on host "postgres" (192.168.224.5) and accepting
TCP/IP connections on port 5432?
', crates/db_schema/src/utils.rs:217:56
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'Error connecting to postgres://lemmy:86ruE60ExlfUQwzAS7e6IGqOB8Z93JkP@postgres:5432/lemmy: could not connect to server: Connection refused
Is the server running on host "postgres" (192.168.224.5) and accepting
TCP/IP connections on port 5432?
', crates/db_schema/src/utils.rs:217:56
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'Error connecting to postgres://lemmy:86ruE60ExlfUQwzAS7e6IGqOB8Z93JkP@postgres:5432/lemmy: could not connect to server: Connection refused
Is the server running on host "postgres" (192.168.224.5) and accepting
TCP/IP connections on port 5432?
', crates/db_schema/src/utils.rs:217:56
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
federation enabled, host is lemmy.d.thewooskeys.com
Starting http server at 0.0.0.0:8536
federation enabled, host is lemmy.d.thewooskeys.com
Starting http server at 0.0.0.0:8536
```

**Reply by EnigmaCurry** [reply]
I think it finally connects, the log you see is from before. if you restart the service (`make restart service=lemmy`) and show the log, it shows just the last two lines repeated again, informing that its up.

I think the lemmy service is starting before postgres is up, and fails, but it has a retry process that finally fixes itself.

**Reply by mcmikemn** [reply]
But I still get "bad gateway". The router and service look error free in Traefik.

**Reply by mcmikemn** [reply]
If I was wrong about the `{{ }}` expressions in `config.hjson` - if they are replaced a la jinja or are read as variables, then if I removed the curly braces and replaced the entire expression with my values, might that be causing this type of error?

Their original file:
```
{
  # for more info about the config, check out the documentation
  # https://join-lemmy.org/docs/en/administration/configuration.html

  database: {
    host: postgres
    password: "{{ postgres_password }}"
  }
  hostname: "{{ domain }}"
  pictrs: {
    url: "http://pictrs:8080/"
    api_key: "{{ postgres_password }}"
  }
  email: {
    smtp_server: "postfix:25"
    smtp_from_address: "noreply@{{ domain }}"
    tls_type: "none"
  }
}
```

**Reply by EnigmaCurry** [reply]
bad gateway is a separate issue, and it may be coming from nginx or it make be coming from traefik, check the logs of each to know for sure.

**Reply by mcmikemn** [reply]
My nginx log never actually says nginx up has started up, only that it's ready to:
```
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
2023/07/30 21:07:14 [notice] 1#1: using the "epoll" event method
2023/07/30 21:07:14 [notice] 1#1: nginx/1.25.1
2023/07/30 21:07:14 [notice] 1#1: built by gcc 12.2.1 20220924 (Alpine 12.2.1_git20220924-r4)
2023/07/30 21:07:14 [notice] 1#1: OS: Linux 5.4.0-153-generic
2023/07/30 21:07:14 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2023/07/30 21:07:14 [notice] 1#1: start worker processes
2023/07/30 21:07:14 [notice] 1#1: start worker process 30
2023/07/30 21:07:14 [notice] 1#1: start worker process 31
```

**Reply by mcmikemn** [reply]
In my Traefik logs, I get only 1 line when I `grep` for "lemmy":
```
$ docker logs traefik-traefik-1 | grep -i lemmy
2023/07/29 12:32:23 reverseproxy.go:667: httputil: ReverseProxy read error during body copy: unexpected EOF
```

**Comment by EnigmaCurry** [comment]
> so I think it has something to do with ports now that both nginx and Traefik are reverse-proxying.

every service listed in the docker-compose is a part of the same docker network, (ie. each docker-compose project gets a `default` network, and unless you specify explicit networks (you haven't) they are all on one network.) So there is no traefik or nginx to worry about, lemmy container *should* be able to connect to postgres container directly, no middleman.

**Comment by EnigmaCurry** [comment]
> they are replaced a la jinja or are read as variables,

config/setup.sh uses `envsubst` which want to replace environment variables eg. ${VARIABLE} - not jinga style {{ VARIABLE }}

**Reply by mcmikemn** [reply]
Yes, but do any of Lemmy's services need them in jinja style?

**Reply by EnigmaCurry** [reply]
I'm not sure, but this is why I like to name templates as such, for example I would name the file `config.template.hjson`

**Reply by EnigmaCurry** [reply]
it did seem to know what it was talking about when it errored tho:

```
thread 'main' panicked at 'Error connecting to postgres://lemmy:86ruE60ExlfUQwzAS7e6IGqOB8Z93JkP@postgres:5432/lemmy: could not connect to server: Connection refused
Is the server running on host "postgres" (192.168.224.5) and accepting
```

this has all the password info from your var, so I think that part is working.

**Reply by mcmikemn** [reply]
Without knowing what they're doing with this file, there's no way to know whether I should replace`password: "{{ postgres_password }}"` with 
- `password: "{{ ${MY_PW_ENV_VAR} }}"`
- `password: "${MY_PW_ENV_VAR}"`
- `password: ${MY_PW_ENV_VAR}`
...or whether I need to leave it as it is and define an env var named `postgres_password`.

**Reply by mcmikemn** [reply]
> this has all the password info from your var, so I think that part is working.

That's a good point.

**Comment by EnigmaCurry** [comment]
@mcmikemn I fixed this and pushed in 37a2b09 - you hadn't written the nginx config file to the right location, so it was running with the default config on port 80. Thats why it was a Bad Gateway.

I replaced the nginx-config container with a custom nginx build with a Dockerfile that adds the static nginx config to /etc/nginx/nginx.conf - since the config is static, it does not need to be run through `envsubst` and, in fact it fails if you do this because there are some variables that are existing that need to stay put `$http_connection` and the like, `envsubst` can be configured to ignore all variables except the ones we want, but then we have to be explicit. Turns out we don't need to because its static and we don't need any variables replaced, so I ripped out the nginx-config  container and replaced it with the nginx Dockerfile that adds the static config. It now starts up correctly on port 8536 and theres no more Bad Gateway.

**Reply by mcmikemn** [reply]
> you hadn't written the nginx config file to the right location

Doh! Where was I saving it to? In both of my `setup.sh` I was trying to write the `envsubst`-ed file to a specific directory, but the files ended up in the root directory of the docker volume (`/var/lib/docker/volumes/<config-dir>/_data/`), so I must be mistunderstanding something there. 

But it doesn't matter because you simplified this by using a static nginx config file. Yay!

**Reply by EnigmaCurry** [reply]
> Where was I saving it to?

you had the right directory just the wrong filename - you had /etc/nginx/nginx-internal.conf instead of /etc/nginx/nginx.conf

**Reply by mcmikemn** [reply]
Ah. I wouldn't have figured that one out, so thanks for the better solution!
